### PR TITLE
ECCV 2020 tutorial: Navigation Basics

### PR DESCRIPTION
**Pull request by bigbike**
## Motivation and Context
This is to finalize the tutorial.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by Skylion007** [comment]
How did you commit just this ipynb? It should have tried to generate the python script automatically with pre-commit @bigbike . If it didn't, you can run the script in tools to run precommit on the specified file.

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py [hunk 1]
```diff
@@ -0,0 +1,1130 @@
...
+except ImportError:
+    import sys, os
+
+    if "google.colab" in sys.modules:
+        # ADDS conda installed libraries to the PYTHONPATH
+        conda_path = "/usr/local/lib/python3.6/site-packages/"
+        user_path = "/root/.local/lib/python3.6/site-packages/"
+        sys.path.insert(0, conda_path)
+        sys.path.insert(0, user_path)
+os.chdir("/content/habitat-sim")
```
Comment (Skylion007) [inline examples/tutorials/nb_python/ECCV_2020_Navigation.py:61]: This should use the `%cd` python magic so it doesn't run locally

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py [hunk 2]
```diff
@@ -0,0 +1,1130 @@
+# ---
```
Comment (Skylion007) [inline examples/tutorials/nb_python/ECCV_2020_Navigation.py:1]: The python script should be added to the testing suite in test_examples

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py [hunk 3]
```diff
@@ -0,0 +1,1130 @@
...
+        "turn_right": habitat_sim.agent.ActionSpec(
+            "turn_right", habitat_sim.agent.ActuationSpec(amount=30.0)
+        ),
+    }
+
+    return habitat_sim.Configuration(sim_cfg, [agent_cfg])
+
+
+# %%
+cfg = make_cfg(sim_settings)
```
Comment (Skylion007) [inline examples/tutorials/nb_python/ECCV_2020_Navigation.py:335]: You can wrap this stuff in a if __name__ == `__main__` if you want, but it's not required.

**Comment by Skylion007** [comment]
The colab will appear at this link until the PR is committed btw: https://colab.research.google.com/github/facebookresearch/habitat-sim/blob/colabs-navigation/examples/tutorials/colabs/ECCV_2020_Navigation.ipynb

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py (hunk 2 above)
Comment (Skylion007) [reply examples/tutorials/nb_python/ECCV_2020_Navigation.py:1]: isort also needs to be rerun

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py [hunk 4]
```diff
@@ -0,0 +1,1130 @@
...
+dir_path = repo.working_tree_dir
+data_path = os.path.join(dir_path, "data")
+# @markdown Optionally configure the save path for video output:
+output_directory = "examples/tutorials/nav_output/"  # @param {type:"string"}
+output_path = os.path.join(dir_path, output_directory)
+if not os.path.exists(output_path):
+    os.mkdir(output_path)
+
+
+def make_video_cv2(observations, prefix="", open_vid=True, fps=60, multi_obs=False):
```
Comment (Skylion007) [inline examples/tutorials/nb_python/ECCV_2020_Navigation.py:865]: This should be in viz_utils now, unless this function is different somehow?

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py [hunk 5]
```diff
@@ -0,0 +1,1130 @@
...
+    navmesh_settings.filter_ledge_spans = True  # @param {type:"boolean"}
+    # default = True
+    navmesh_settings.filter_walkable_low_height_spans = True  # @param {type:"boolean"}
+    # default = True
+
+    # @markdown **Detail mesh generation parameters**:
+    # @markdown For more details on the effects
+    navmesh_settings.region_min_size = (
+        20
+    )  # @param {type:"slider", min:0, max:50, step:1}
```
Comment (Skylion007) [inline examples/tutorials/nb_python/ECCV_2020_Navigation.py:793]: Black broke all these sliders... hmm... we may need to run off the fmt manually with ```# fmt : off```

see black's documentation. NB: @aclegg3

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py (hunk 4 above)
Comment (aclegg3) [reply examples/tutorials/nb_python/ECCV_2020_Navigation.py:867]: We're targeting more work on this soon. 👍

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py (hunk 5 above)
Comment (Skylion007) [reply examples/tutorials/nb_python/ECCV_2020_Navigation.py:797]: If this doesn't work, we may have to build a quick shim with ipython widgets.

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py [hunk 6]
```diff
@@ -0,0 +1,1130 @@
...
+#
+#   *Decrease these for better accuracy at the cost of higher compute cost.*
+#
+#   **Note:** most continuous parameters are converted to multiples of cell dimensions, so these should be compatible values for best accuracy.
+#   - **cell_size** - xz-plane voxel dimensions. [Limit: >= 0]
+#   - **cell_height** - y-axis voxel dimension. [Limit: >= 0]
+#
+# - **Agent parameters**:
+#
+#   - **agent_height** - Hieght of the agent. Used to cull navigable cells with obstructions.
```
Comment (Skylion007) [inline examples/tutorials/nb_python/ECCV_2020_Navigation.py:722]:
```suggestion
#   - **agent_height** - Height of the agent. Used to cull navigable cells with obstructions.
```

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py (hunk 5 above)
Comment (bigbike) [reply examples/tutorials/nb_python/ECCV_2020_Navigation.py:797]: This part belongs to @aclegg3 . He may need to take a look.

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py (hunk 4 above)
Comment (jturner65) [reply examples/tutorials/nb_python/ECCV_2020_Navigation.py:867]: Yeah, sorta got held up with finishing up scene import.  We're working on it :)

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py (hunk 5 above)
Comment (jturner65) [reply examples/tutorials/nb_python/ECCV_2020_Navigation.py:797]: I thought we were trying to avoid the resultant python files in black anyway, because of the colab decorators that get moved around.

Comment (Skylion007) [reply examples/tutorials/nb_python/ECCV_2020_Navigation.py:797]: Yeah, those sections should be marked off with `#fmt: off/on` for that reason @jturner65

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py [hunk 7]
```diff
@@ -45,7 +45,22 @@
 # @title Colab Setup and Imports { display-mode: "form" }
 # @markdown (double click to see the code)
 
-import os, sys
+import math
```
Comment (Skylion007) [inline examples/tutorials/nb_python/ECCV_2020_Navigation.py:48]: Weird, why didn't the ipynb file update. Pre-commit runs automatically when you commit, right @bigbike ?

File: tests/test_examples.py [hunk 8]
```diff
@@ -55,6 +55,7 @@ def run_main_subproc(args):
             "--no-show-video",
             "--no-make-video",
         ),
+        ("examples/tutorials/nb_python/ECCV_2020_Navigation.py", "--no-show-images"),
```
Comment (Skylion007) [inline tests/test_examples.py:58]: Is the --no-show-images arg even parsed anywhere by the script? If not, you can just drop that arg.

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py (hunk 7 above)
Comment (bigbike) [reply examples/tutorials/nb_python/ECCV_2020_Navigation.py:48]: I have no idea at this point. 

I even upgraded my git to the latest version. 
If I changed the .py file, ipynb will not be updated.

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py [hunk 9]
```diff
@@ -53,30 +53,28 @@
...
+# %matplotlib inline
+from matplotlib import pyplot as plt
 # function to display the topdown map
 from PIL import Image
 
-import habitat_sim.utils.common as utils
+from habitat_sim.utils import common as utils
 from habitat_sim.utils import viz_utils as vut
 
 try:
```
Comment (Skylion007) [inline examples/tutorials/nb_python/ECCV_2020_Navigation.py:65]: This fancy import pattern is no longer needed btw. You can just do import habitat_sim now.

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py [hunk 10]
```diff
@@ -0,0 +1,1105 @@
...
+    total_frames += 1
+
+
+# %% [markdown]
+# # Working with the NavMesh
+
+# %% [markdown]
+# Habitat-sim provides pathfinding and navigability constraints via intergration with [Recast Navigation | Detour](https://masagroup.github.io/recastdetour/) through the [nav module](https://aihabitat.org/docs/habitat-sim/habitat_sim.nav.html).
+#
+# This tutorial section demonstrates loading/recomputing/saving a NavMesh for a static scene and using it for a discrete navigation task. This module can also be applied to continuous navigation tasks, which is covered in the [advanced features tutorial (TODO: link this)](https://).
```
Comment (Skylion007) [inline examples/tutorials/nb_python/ECCV_2020_Navigation.py:421]: TODO: Add the link:

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py [hunk 11]
```diff
@@ -0,0 +1,1094 @@
...
+    bounds = pathfinder.get_bounds()
+    for point in points:
+        # convert 3D x,z to topdown x,y
+        px = (point[0] - bounds[0][0]) / meters_per_pixel
+        py = (point[2] - bounds[0][2]) / meters_per_pixel
+        points_topdown.append(np.array([px, py]))
+    return points_topdown
+
+
+def display_topdown_map(topdown_map, trajectory=None, key_points=None):
```
Comment (Skylion007) [inline examples/tutorials/nb_python/ECCV_2020_Navigation.py:454]: We have a unified utility for this in a upcoming PR too

File: examples/tutorials/nb_python/ECCV_2020_Navigation.py [hunk 12]
```diff
@@ -0,0 +1,1094 @@
...
+
+    if not custom_height:
+        height = scene_bb.y().min
+
+    # This map is a 2D boolean array
+    topdown_map = sim.pathfinder.get_topdown_view(meters_per_pixel, height)
+    display_topdown_map(topdown_map)
+
+except:
+    print("Error - aborting")
```
Comment (Skylion007) [inline examples/tutorials/nb_python/ECCV_2020_Navigation.py:521]: You may want to print the error traceback here using the traceback module. That way the user can debug

**Comment by Skylion007** [comment]
So I tested this at the link I posted above and it works! Feel free to give it a try.

**Comment by Skylion007** [comment]
@aclegg3, currently Habitat-API is installed after habitat-sim is in the .circleci build. That's why the test is failing. We need to either handle that import failing or install Habitat-API earlier

File: examples/tutorials/colabs/ECCV_2020_Navigation.ipynb [hunk 13]
```diff
@@ -140,7 +135,12 @@
     "else:\n",
     "    show_video = False\n",
     "    make_video = False\n",
-    "    display = False"
+    "    display = False\n",
+    "\n",
+    "# import the maps module alone for topdown mapping\n",
+    "if display:\n",
+    "    sys.path.insert(1, \"/content/habitat-api/\")\n",
```
Comment (Skylion007) [inline examples/tutorials/colabs/ECCV_2020_Navigation.ipynb:142]: Why is this necessary? Habitat-API should already be in the PATH in the system wide install?

Comment (Skylion007) [reply examples/tutorials/colabs/ECCV_2020_Navigation.ipynb:142]: The bug was solved by #731

**Comment by Skylion007** [comment]
@aclegg3 Still two minor issues whenever you get the chance, fix the unnecessary import shim and redo to use the new make_video function. :)
